cpp/src/io/json/write_json.cu

```diff
@@ -0,0 +1,743 @@
+{
+  std::vector<std::string> escaped_column_names;
+  std::transform(column_names.begin(),
+                 column_names.end(),
+                 std::back_inserter(escaped_column_names),
+                 [](column_name_info const& name_info) {
+                   std::string name = name_info.name;
+                   char const quote = '\"';
+                   if (name.empty() ||           // no header name
+                       name.front() == quote) {  // header already quoted
+                     return name;
+                   }
+
+                   // escape any quotes
+                   size_t pos = 0;
+                   while ((pos = name.find(quote, pos)) != name.npos) {
+                     name.insert(pos, 1, quote);
+                     pos += 2;
+                   }
+
+                   // check if overall quotes are required
+                   name.insert(name.begin(), quote);
+                   name.insert(name.end(), quote);
+                   return name;
+                 });
+  if (escaped_column_names.empty()) {
+    std::generate_n(std::back_inserter(escaped_column_names), num_columns, [v = 0]() mutable {
+      return std::to_string(v++);
+    });
+  }
+  return make_strings_column_from_host(escaped_column_names, stream);
+}
+
+void write_chunked(data_sink* out_sink,
+                   strings_column_view const& str_column_view,
+                   std::string const& line_terminator,
+                   json_writer_options const& options,
+                   rmm::cuda_stream_view stream,
+                   rmm::mr::device_memory_resource* mr)
+{
+  CUDF_EXPECTS(str_column_view.size() > 0, "Unexpected empty strings column.");
+
+  string_scalar d_line_terminator{line_terminator};
+  auto p_str_col_w_nl = cudf::strings::detail::join_strings(str_column_view,
+                                                            d_line_terminator,
+                                                            string_scalar("", false),
+                                                            stream,
+                                                            rmm::mr::get_current_device_resource());
+  strings_column_view strings_column{p_str_col_w_nl->view()};
+
+  auto total_num_bytes      = strings_column.chars_size();
+  char const* ptr_all_bytes = strings_column.chars_begin();
+
+  if (out_sink->is_device_write_preferred(total_num_bytes)) {
+    // Direct write from device memory
+    out_sink->device_write(ptr_all_bytes, total_num_bytes, stream);
+  } else {
+    // copy the bytes to host to write them out
+    thrust::host_vector<char> h_bytes(total_num_bytes);
+    CUDF_CUDA_TRY(cudaMemcpyAsync(h_bytes.data(),
+                                  ptr_all_bytes,
+                                  total_num_bytes * sizeof(char),
+                                  cudaMemcpyDeviceToHost,
+                                  stream.value()));
+    stream.synchronize();
```

Comment:
Also could be created using the vector factory (_sync version).
I realize this is a copy of existing code that we need to update at some point 😅

Reply:
used sync version.